Compute bounding boxes approximating cells with a property (adding in grid_tools)

GivAlz opened the pull request:
Adding the following function to distributed/grid_tools
```cpp
    template < int dim, int spacedim>
    std::vector<BoundingBox<spacedim> >
    compute_locally_owned_bounding_box
    (const parallel::distributed::Triangulation< dim, spacedim > &distributed_tria)
```
The function returns a vector of BoundingBoxes (not less than two, usually no more than 3...but it really depends on the triangulation and dimension) approximating the locally_owned cells.

The aim is to describe connected components and other "important" shape features and then communicate them to other processes. While these bounding boxes are allowed to cover more space, every locally_owned cell must be inside one of them.

I believe the basic idea of the function is simple: go to the coarsest level at which you can distinguish connected components (which is the biggest thing we want to catch) and build bounding_boxes at this level.

I worked on my own to build this function: comments, suggestions etc are all very welcome!
I hope it's not too bad, but I really don't know because I couldn't find much material to start from.

I believe a better know-how of p4est could help (skipping some cells, for instance...is it possible to know if all children cells of the current one are locally_owned?) and there are some other things that could be changed (for instance merging cells if the union covers the same space as the two original ones).

Since this function is part of a project I'm working at with @luca-heltai  and @gassmoeller I'm hoping to create, asap, a working example to benchmark it and check where the bottlenecks are.

Note: I also corrected two small things in Bounding Box class: a documentation error and a forgotten return at the end of the function...

davydden commented on include/deal.II/base/grid_tools.h:
```diff
@@ -0,0 +1,429 @@
+
+    template < int dim, int spacedim>
+    std::vector<BoundingBox<spacedim>>
+                                    compute_locally_owned_bounding_box
```
it looks like you need to run `astyle` indentation script

bangerth replied:
Nah, this is just how astyle does it because it doesn't understand (yet) that `>>` is not an operator but two closing template arguments. You can get nicer indentation if you write it as `> >` (with a space).

bangerth commented on include/deal.II/base/grid_tools.h:
```diff
@@ -0,0 +1,429 @@
+// ---------------------------------------------------------------------
```
This can't be right -- you've moved the whole file into the `base/` directory. It should stay in `distributed/`.

bangerth commented on include/deal.II/distributed/grid_tools.h:
```diff
@@ -130,6 +131,21 @@ namespace parallel
+     * which has been ordered following the Morton-Z curve order.
+     */
+
+    template < int dim, int spacedim>
```
no need for the empty line

bangerth commented on include/deal.II/distributed/grid_tools.h:
```diff
@@ -130,6 +131,21 @@ namespace parallel
+    /**
+     * Compute a vector of bounding boxes which approximates the space occupied by
+     * the locally_owned cells. The number of bounding boxes (two or more) tries
+     * to catch the general shape of the space occupied by the locally_owned_cells,
```
in both lines, no need for the underscores in `locally_owned(_cells)`

bangerth commented on include/deal.II/distributed/grid_tools.h:
```diff
@@ -130,6 +131,21 @@ namespace parallel
                                   std::function<void (const typename MeshType::active_cell_iterator &, const DataType &)> unpack);
+
+    /**
+     * Compute a vector of bounding boxes which approximates the space occupied by
```
I think you want to say "encloses" instead of "approximates". The latter suggests that the inclusion is not exact, but then it would be useless.

bangerth commented on include/deal.II/distributed/grid_tools.h:
```diff
@@ -130,6 +131,21 @@ namespace parallel
+     * describing connected components and bigger shape variations.
+     *
+     * This function is meant to work only with a parallel::distributed::Triangulation
+     * which has been ordered following the Morton-Z curve order.
```
Is this important? I'd think it can be used for any triangulation, regardless of how exactly the cells are subdivided.

bangerth commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,166 @@
+      lastc = distributed_tria.last_active();
+
+      while ( ! firstc->is_locally_owned() )
+        ++firstc;
```
This (and the next block) isn't correct if a processor has no locally owned cells -- you'd eventually run into an assertion because you're trying to increment an iterator that is past-the-end.

tjhei commented on tests/distributed_grids/grid_tools_compute_bounding_box_1.cc:
```diff
@@ -0,0 +1,95 @@
+
+#include <vector>
+
+//#include "/home/giovy/opt/src/dealii_fork/include/deal.II/distributed/grid_tools.h"
```
remove

bangerth commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,166 @@
+            }
+          std::vector<BoundingBox<spacedim>> vBBox;
+          vBBox.push_back(BoundingBox<spacedim>(std::make_pair( minp, maxp )));
+          return vBBox;
```
I suspect that you could simplify this to
```
  return std::vector<BoundingBox<spacedim>>({BoundingBox<spacedim>(std::make_pair( minp, maxp ))});
```

bangerth commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,166 @@
+        --lastc;
+
+      //Degenerate cases: only one owned cell or we're at level 0
+      if (lastc==firstc || firstc->level()*lastc->level()==0)
```
This is overly complicated to read -- just do this (assuming that's what you want):
```
  if (lastc==firstc || firstc->level()==0 || lastc->level()==0)
```

tjhei replied:
I don't understand the || and the second condition. Doesn't that mean that it is enough that one of the cells is at level 0? Why do you need to consider only the cell firstc if the first expression is false?

bangerth commented on tests/base/bounding_box_1.cc:
```diff
@@ -134,4 +134,6 @@ int main()
   deallog << std::endl << "Test for dimension 3, unitary box" << std::endl;
   test_unitary();
+
+  return 0;
```
I don't think you need this.

bangerth commented on tests/distributed_grids/grid_tools_compute_bounding_box_1.cc:
```diff
@@ -0,0 +1,95 @@
+  test_hypercube<2> ();
+  test_hypercube<3> ();
+
+  return 0;
```
no need -- it is implied

tjhei commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,166 @@
+      // we need to handle the possibility of a second connected component.
+      //
+      // Assuming the triangulation is distributed following the morton-z curve order
+      // we know that, if there are two connected components, then the first and the
```
please assert that all locally owned cells are exactly within [firstc, lastc] in debug mode. I am not sure this is correct, by the way. We do not iterate over cells in morton ordering.

tjhei commented on include/deal.II/distributed/grid_tools.h:
```diff
@@ -130,6 +131,21 @@ namespace parallel
+     * the locally_owned cells. The number of bounding boxes (two or more) tries
+     * to catch the general shape of the space occupied by the locally_owned_cells,
+     * describing connected components and bigger shape variations.
+     *
```
I assume your bounding box only contains the vertices (which does not handle curved geometry). Can you leave a comment here?

tjhei commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,166 @@
+      while (parentc->level()>0)
+        {
+          parentc = tmpc->parent();
+          if (parentc->point_inside(lastc_center))
```
I am not sure if this is robust enough (I don't quite understand the logic). There are cases where the center of a child cell is not inside all parents (think refining curved geometry).

tjhei commented on include/deal.II/base/grid_tools.h:
```diff
@@ -0,0 +1,429 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
```
why is this file in base/ ? Did you add it by accident?

GivAlz commented:
Thank you everyone for the comments! Sorry for all the avoidable mistakes I made (copying files around by mistake, leaving useless comments etc.). I tried to face the various questions by deeply changing the function.

Now it relies much less on the ordering: to get the first parent cell of both firstc and lastc I use the parent function (similarly to what I've seen done in other deal.ii functions). So thank you, @tjhei , for pointing out all the problems in robustness: I hope the new version has improved.

I want to thanks @bangerth : thank you for that code! What I'm trying to do is to catch with as little bounding boxes as possible as much of the shape as possible. My algorithm won't do that, but if the order is z-morton (or close enough) it should be able to describe accurately the following (each number is a cell and describes its owner):
0 0 1 1 1
0 0 0 1 1
This because often the "Locally owned domain" is almost entirely contained in a coarse cell and then, attached to it, there is a small "lump" of owned cells, similar to this:
![distributed](https://user-images.githubusercontent.com/9434511/30695311-043a4486-9ed8-11e7-89b6-58005a0d33f1.png)
I realize my way of choosing the parent cell is quite arbitrary: I believe testing it with different meshes and comparing it to your method is a good idea.

Also: currently in some cases (see the example with 2 processes) it would be better to merge the bounding boxes; I don't know if it's better to add this "merging routine" or just use more bounding boxes.

tjhei commented:
> I realize my way of choosing the parent cell is quite arbitrary

Yes. You could describe everything with a single BB, you could just return a BB for every locally owned cell, or you can do anything in-between (merging bounding boxes from higher level cells). What we really have is a AABB tree (axis-aligned bounding box tree).

Would it make more sense to supply an additional argument to your function that determines the level up to which you merge the bounding boxes of children? (level=0 would return a single bounding box, level=infinity would return a vector of cell bounding boxes, etc.). Alternatively, you could generate the whole AABB tree instead and let the user decide what to do with it.

I don't know how you are planning to use this functionality, but for queries like "is this point within my part of the mesh?" having the tree is probably quite useful.

bangerth commented:
As a matter of procedure, it would probably make your code easier to read if you had functions `TriaAccessor::get_bounding_box()` and `BoundingBox::merge_with(const BoundingBox &)`. These are generically useful operations and they will make the code you want to write here easier to read and write.

Feel free to propose such functions in separate patches. I bet they would quickly be approved because their implementation is pretty uncontroversial.

GivAlz commented:
The aim is to pass around the bounding box so that each core has an "idea" of the space owned by others (thus, if has a second mesh withing that area, it can send the necessary information to the processes it guesses own that domain part).
Clearly the aim is to have as little BoundingBoxes as possible while describing the space with a good enough accuracy. I don't know why I didn't think about setting a level for the bounding boxes: I shall implement the functions to merge/get_bounding_box while thinking about how to implement such an option (which seems quite reasonable).
Thank you all very much!!!

bangerth commented:
Yes, having one bounding box per cell is not very useful -- in that case one might as well work with the cells right away :-)

Rombur commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,184 @@
+        }
+      else
+        {
+          for (auto cell: distributed_tria.cell_iterators_on_level(0))
```
So the code below is exactly the same as the code above but the range of the loop is different? It would be nice if you encapsulate the code in one function instead of copy/pasting it (double the code, double the bugs).

davydden replied:
Indeed, you can refactor the same parts into internal function

GivAlz replied:
Ok, in the end I used another function.

Rombur commented on include/deal.II/distributed/grid_tools.h:
```diff
@@ -130,6 +131,17 @@ namespace parallel
+
+    /**
+     * Compute a vector of bounding boxes which encloses the space occupied by
+     * the locally owned cells. The number of bounding boxes (two or more) tries
```
So even if the partitioning is such that one bounding box will perfectly describe the domain, I will get two of them?

GivAlz replied:
I added a new version where there is a flag to allow or not merging.
If it is not allowed then yes, two bounding boxes shall be returned.
If it is allowed then the two bounding boxes shall be merged if they can be described as one (for instance (0,0) (1,1) and (1,0), (2,1) become (0,0), (2,1)) ).

bangerth commented:
@GivAlz -- if you want to add the picture that @davydden mentioned: Just put it into `doc/doxygen/images` and then link to it using `@image html my_picture.png` in the documentation of the function or class where you want to use it. You can take a look at several examples in `include/deal.II/grid/grid_generator.h` to see how this is done.

davydden commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,184 @@
+    template < int dim, int spacedim>
+    std::vector<BoundingBox<spacedim> >
+    compute_locally_owned_bounding_box
+    (const parallel::distributed::Triangulation< dim, spacedim > &distributed_tria)
```
I don’t think there is anything specific to p::d::tria, u can accept p::tria

GivAlz replied:
Yes, this function (also the one I added "compute_cell_locally_owned_bounding_box") could work with any kind of triangulation but it would make no sense since, in that case, everything is locally owned or, if it is shared, we don't need a bounding box for a "general" description of the space since we already have it all.

Do you think it'd be better anyway to use a general MeshType?

davydden replied:
> would make no sense since, in that case, everything is locally owned

nope. First, you have parallel::shared::Triangulation, second, people can be using Triangulation but partitioning it manually. So what you really want to do here is: a) take standard Triangulation b) Add arbitrary predicate.

davydden replied:
> if it is shared, we don't need a bounding box for a "general" description of the space since we already have it all.

why is that? You don't know how tria is partitioned by Metis, it would still be useful to have a simple representation of the domain specified by a predicate (i.e. locally owned) even for p::s::Tria.

davydden commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,184 @@
+      typename parallel::distributed::Triangulation< dim, spacedim >::active_cell_iterator
+      lastc = distributed_tria.last_active();
+
+      while ( ! (firstc->is_locally_owned() || firstc == lastc ) )
```
I think the algorithm is general enough to accept a predicate instead of hard-coding locally owned cells

GivAlz replied:
Ok, then it would make sense to have it not only for distributed meshes...I see where this is going. Thank you!

davydden commented on source/distributed/CMakeLists.txt:
```diff
@@ -16,6 +16,7 @@
 INCLUDE_DIRECTORIES(BEFORE ${CMAKE_CURRENT_BINARY_DIR})
 
 SET(_unity_include_src
+  grid_tools.cc
```
if you don't need `p::d::Tria`, please put the function inside the already existing `soruce/grid/grid_tools.cc`.

davydden commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,184 @@
+                        for ( unsigned int d=0; d<spacedim; ++d)
+                          {
+                            minp[d] = std::min( minp[d], local_cells[i]->vertex(v)[d]);
+                            maxp[d] = std::max( maxp[d], local_cells[i]->vertex(v)[d]);
```
> Your bounding box won't work if you have a curved edge 

exactly @tjhei 's comment https://github.com/dealii/dealii/pull/5117#discussion_r140021562 which probably got hidden by some changes.

GivAlz replied:
The if is there just to initialize the bounding box, yes. I'm trying to make it work with max/-max the problem is that this must be done with a vertex of a locally owned cell, right?

As for the curved geometry: the problem is that, in those cases, edges are curved right? Thank you!

GivAlz commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,184 @@
+      //enclosing all the locally owned cells.
+      //The idea is to find a group of parent cells which encloses all the locally
+      //owned cells and, for each of them, produce a bounding box of the locally owned
+      //points inside it.
```
Yes, that doesn't make much sense. I meant the vertices of locally owned cells

GivAlz commented:
Ok, so I worked on this but I didn't follow well your comments, sorry. Today I hope to answer them all/update the code.

I guess you already saw the other pull request, which I closed, anyway now I'm thinking of this method which gets rid of the arbitrary way in choosing the parent cell (by actually letting the user choose).

I've added a function for bounding boxes which returns 0 if they're not neighbours, 1 or 2 otherwise (to help with the following merging idea).

I added two function to distributed triangulation

    one which, given a cell, creates the bounding box of the locally owned cells inside it (this is done in the dumbest way possible)
    another which, given a refinement level and a target number of boxes, uses the previous function on cells of the given refinement level (or less, if they're active) and then merges the results: first by merging cells which can be expressed with a single bounding box, then by merging looking for the smallest one to a neighbour (not the brightest idea, I know...)

I want to thank you for all the comments in the other thread. This is still a work in progress but it is clear I need to update more often here for comments (and checking if other people are working on bounding boxes).

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -77,16 +77,35 @@ class BoundingBox
+   *  V--W--.    V-----.
+   *  @endcode
+   */
+  unsigned int is_neighbour(const BoundingBox<spacedim,Number> &other_bbox, const double &err=1e-12) const;
```
why do you need `err` as a parameter? You could as well used internally
```
100. * std::numeric_limits<double>::epsilon
```
or alike.

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -77,16 +77,35 @@ class BoundingBox
+   * @p err, false otherwise.
    */
-  bool point_inside (const Point<spacedim, Number> &p) const;
+  bool point_inside (const Point<spacedim, Number> &p, const double &err=1e-12) const;
```
same here

GivAlz replied:
Ok, I didn't know about the internal parameter. I shall revert the functions back in the next commit. Thank you!

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,58 @@ namespace GridTools
+  template < class MeshType >
+  BoundingBox < MeshType::space_dimension >
+  compute_cell_predicate_bounding_box
+  (const typename MeshType::cell_iterator &parent_cell,
```
this function should take
```
( const MeshType                                                                    &mesh,
  const std::function<bool (const typename MeshType::active_cell_iterator &)> &predicate );
```

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,58 @@ namespace GridTools
+   */
+  template < class MeshType >
+  BoundingBox < MeshType::space_dimension >
+  compute_cell_predicate_bounding_box
```
do you need this function in public domain (available to the user)? I would say no, so I would put it into your `.cc`

GivAlz replied:
I agree it's highly unlikely someone will want to use this function by itself, that includes me of course! The current version of compute_mesh_predicate_bounding_box uses it: I created it to make the code more readable. If there's a better way to encapsulate it, or there's some way to hide it from the public, I'm willing to do it.

Also if there was a function such as "cell->is_children(parent)" which worked between different layers I could think about using the already present compute_bounding_box inside compute_mesh_predicate_bounding_box but I couldn't find one...do you know if I'm missing something?

davydden replied:
> If there's a better way to encapsulate it, or there's some way to hide it from the public

that's exactly what I already suggested above :smile:  declare and use it from your `grid_tools.cc` file, say
```
namespace internal
{
    //blabla
}
```

> do you know if I'm missing something?

that i don't know

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,58 @@ namespace GridTools
+  compute_mesh_predicate_bounding_box
+  ( const MeshType &mesh,
+    const std::function<bool (const typename MeshType::active_cell_iterator &)> &predicate,
+    const unsigned int &refinement_level, const unsigned int &max_boxes, const double &err = 1e-12);
```
you can set default to `refinement_level` being the coarsest level in `mesh`.
For `max_boxes` set to `numbers::invalid_unsigned_int` (and use as default) to return all boxes. Add a check that the number is positive.
Finally, you don't need `err` here either

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,58 @@ namespace GridTools
+   * - @p predicate : the property of the cells to enclose (see compute_cell_predicate_bounding_box)
+   * - @p refinement_level : it describes the level at which the initial Bounding Boxes are created. The refinement
+   *  should be set to a low refinement level. If @p refinement_level is higher than the number of levels of the
+   *  triangulation or then an exception is trown.
```
something is missing after `or`

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,58 @@ namespace GridTools
+   * - @p max_boxes : the maximum number of bounding boxes to compute. If more are created the smaller ones are
+   *  merged with neighbours. If the value is set to 0 all computed bounding boxes are returned.
+   *  Notice only neighbouring cells are merged (see the @p is_neighbour function in bounding box class): if
+   *  the target number of bounding boxes max_boxes can't be reached by merging neighbours an exception is trown
```
thrown

davydden commented on source/distributed/CMakeLists.txt:
```diff
@@ -16,7 +16,6 @@
 INCLUDE_DIRECTORIES(BEFORE ${CMAKE_CURRENT_BINARY_DIR})
 
 SET(_unity_include_src
-  grid_refinement.cc
```
you don't want to do this

davydden commented on source/distributed/CMakeLists.txt:
```diff
@@ -37,7 +36,6 @@ SETUP_SOURCE_LIST("${_unity_include_src}"
   )
 
 SET(_inst
-  grid_refinement.inst.in
```
same

davydden commented on source/grid/grid_tools.cc:
```diff
@@ -1863,7 +1863,6 @@ namespace GridTools
   }
 
 
-
```
keep the empty line (we usually have 3 empty lines in .cc)

davydden commented:
@GivAlz please run indentation script with `astyle@2.04`

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -77,14 +77,34 @@ class BoundingBox
+   * have dimension spacedim, check if their insersection is non empty
+   *
+   * Return values:
+   * - 0 : not neighbours
```
instead of assigning meaning to each case, use enumerator with those options. Make sure you have tests which cover all 4 cases.

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,33 @@ namespace GridTools
+   * be expressed as a single one. In a last optional step all smaller boxes are merged with neighbours.
+   *
+   * The parameters to control the algorithm are:
+   * - @p predicate : the property of the cells to enclose. See this function's test for an example.
```
refereeing to tests is not useful here, just mention that 

> An example of predicate is IteratorFilters::LocallyOwnedCell .

or alike

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,33 @@ namespace GridTools
+   * The parameters to control the algorithm are:
+   * - @p predicate : the property of the cells to enclose. See this function's test for an example.
+   *  The predicate is tested only on active cells.
+   * - @p refinement_level : it describes the level at which the initial Bounding Boxes are created. The refinement
```
~~it describes~~ defines

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,33 @@ namespace GridTools
+   *  Notice only neighbouring cells are merged (see the @p is_neighbour function in bounding box class): if
+   *  the target number of bounding boxes max_boxes can't be reached by merging neighbours an exception is thrown
+   *
+   * Notice this function doesn't work with curved meshes.
```
@note This function does not take into account curvature of cells, i.e. it's assumed that the mapping is linear.

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1033,33 @@ namespace GridTools
+   *  should be set to a low refinement level. If @p refinement_level is higher than the number of levels of the
+   *  triangulation an exception is thrown.
+   * - @p max_boxes : the maximum number of bounding boxes to compute. If more are created the smaller ones are
+   *  merged with neighbours. By default all boxes are kept after merging and erasing the reduntant ones.
```
> merged with neighbours

Out of curiosity and without looking into implementation, doesn't this lead to graph colouring problem where each box is a node and edges connect boxes which intersect? 

Say if you have some tens of boxes where some overlap, how do you decide which to merge and more importantly in what order? You can, of course, start with boxes which can be grouped into a single one without adding extra volume (the last example in `is_neighbour`), but the rest is still an open question, is it?

GivAlz commented:
Hi everyone!
  This weekend I tried to follow all the comments: with the exception of the picture, ( sorry @bangerth  , but I shall add a better one), but I tried to do my best for each of them.

What's new: now the is_neighbour algorithm is actually readable. I lost so much time thinking about the wrong approach, now it is quite straightforward:
- check if the mesh touch or not
- if they touch compute the intersection: if it has dimension spacedim-1 return
- if it is 3 dimensional check if one is inside the other, otherwise return another type
- check for the vertices: to see if the cells are "mergeable"

I defined an enum as suggested (maybe some names are not extremely bright, acceppting suggestions) @davydden .

I tried initializing directly, as @Rombur suggested, the bounding boxes but I couldn't do better than what I've done so far.

@tjhei the bounding boxes should represent the locally owned cells: I don't want to have many of them for each process. Probably a tree is a good suggestion of how to organize the bounding boxes of all the processes.

Thank you all for your help: while it is sometimes difficult to have code under such scrutiny I really I'm learning a lot :)

bangerth commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,34 @@ namespace GridTools
+   *  The predicate is tested only on active cells.
+   * - @p refinement_level : it defines the level at which the initial Bounding Boxes are created. The refinement
+   *  should be set to a low refinement level. If @p refinement_level is higher than the number of levels of the
+   *  triangulation an exception is thrown.
```
Same as above -- it's not clear to me what "low" actually means for a refinement level -- small (fine-level) cells or large (coarse-level) cells?

bangerth commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,34 @@ namespace GridTools
+   * The parameters to control the algorithm are:
+   * - @p predicate : the property of the cells to enclose e.g. IteratorFilters::LocallyOwnedCell .
+   *  The predicate is tested only on active cells.
+   * - @p refinement_level : it defines the level at which the initial Bounding Boxes are created. The refinement
```
-> bounding box

bangerth commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,34 @@ namespace GridTools
+   *  should be set to a low refinement level. If @p refinement_level is higher than the number of levels of the
+   *  triangulation an exception is thrown.
+   * - @p max_boxes : the maximum number of bounding boxes to compute. If more are created the smaller ones are
+   *  merged with neighbours. By default all boxes are kept after merging and erasing the reduntant ones.
```
I don't know what this means: "all boxes are kept after merging and erasing". Do you erase them, or do you keep them?

reduntant -> redundant

bangerth commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,34 @@ namespace GridTools
+   *  triangulation an exception is thrown.
+   * - @p max_boxes : the maximum number of bounding boxes to compute. If more are created the smaller ones are
+   *  merged with neighbours. By default all boxes are kept after merging and erasing the reduntant ones.
+   *  Notice only neighbouring cells are merged (see the @p is_neighbour function in bounding box class): if
```
write this as "see the BoundingBox::is_neighbor() function" -- this way, doxygen will automatically link to it.

bangerth commented on source/base/bounding_box.cc:
```diff
@@ -25,7 +24,8 @@ bool BoundingBox<spacedim,Number>::point_inside (const Point<spacedim, Number> &
       //The bounding box is defined as a closed set
-      if ( p[i] < this->boundary_points.first[i] || this->boundary_points.second[i] < p[i])
+      if ( p[i] + std::numeric_limits<double>::epsilon() < this->boundary_points.first[i]
+           || this->boundary_points.second[i] + std::numeric_limits<double>::epsilon() < p[i])
```
I think you want to use `std::numeric_limits<Number>` instead.

But this is the wrong formula: http://en.cppreference.com/w/cpp/types/numeric_limits/epsilon shows that epsilon is only useful *when you compare with objects of size 1*. If, however, you are doing quantum mechanics and all of your lengths are on the order of 1e-15, then epsilon is a *very large* tolerance!

I think you want to do something like `p[i] * (1. + std::numeric_limits<Number>::epsilon())` which has the correct length.

bangerth commented on source/base/bounding_box.cc:
```diff
@@ -42,6 +42,83 @@ void BoundingBox<spacedim,Number>::merge_with(const BoundingBox<spacedim,Number>
+      // Step 1: testing if the boxes are close enough to intersect
+      for ( unsigned int d=0 ; d<spacedim; ++d )
+        if ( bbox1[0][d] - std::numeric_limits<double>::epsilon() > bbox2[1][d]
+             || bbox2[0][d] - std::numeric_limits<double>::epsilon() > bbox1[1][d] )
```
same issues here

bangerth commented on source/base/bounding_box.cc:
```diff
@@ -42,6 +42,83 @@ void BoundingBox<spacedim,Number>::merge_with(const BoundingBox<spacedim,Number>
+
+      unsigned int intersect_dim = spacedim;
+      for (unsigned int d=0; d < spacedim; ++d)
+        if ( std::abs( intersect_bbox_min[d] - intersect_bbox_max[d] ) < std::numeric_limits<double>::epsilon() )
```
A similar issue arises here: you want the epsilon to be *relative* to the sizes of the objects. You can achieve this by testing
```
  std::abs( intersect_bbox_min[d] - intersect_bbox_max[d] ) 
  < 
  std::numeric_limits<double>::epsilon() * (std::abs( intersect_bbox_min[d]) + std::abs(intersect_bbox_max[d] ))
```

What happens if `intersect_bbox_min[d] == intersect_bbox_max[d]`? Would it be better to use `<=` instead of `<`?

bangerth commented on source/base/bounding_box.cc:
```diff
@@ -42,6 +42,83 @@ void BoundingBox<spacedim,Number>::merge_with(const BoundingBox<spacedim,Number>
+      bool same_direction = true;
+      for (unsigned int d = 0; d < spacedim; ++d)
+        {
+          if (std::abs(bbox2[0][d] - bbox1[0][d]) > std::numeric_limits<double>::epsilon() )
```
same issue here

GivAlz commented:
Thank you, I changed the various errors regarding relative error and the spelling thing.
I also wrote coarser instead of low all around, sorry for that. As for the merging_function it actually works always, I added that comment in a rush and was meant for "is_neighbour" before I changed the algorithm. Now what happens if one tries to intersect with a degenerate bbox is getting a value "neighbours" (i.e. the intersection has dimension at most spacedim-2...) which is ok.

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,35 @@ namespace GridTools
+   *
+   * The algorithm loops over all cells at given level @p refinement_level and active cells at coarser refinement
+   * levels, creating a Bounding Box for each of them. Then it merges the boxes which can be expressed as a single
+   * one. In a last optional step all smaller boxes are merged with neighbors.
```
I don't understand the last sentence, can you reformulate or expand the description? What are "smaller boxes"? Do you imply that a "merge" is a combination without error? You also seem to describe the algorithm to get the result and not the result in an abstract way. It is typically more helpful for a user to get a description of the output independent on how you achieve it. Maybe something like "a cover of the cells with the minimum number of non-overlapping bounding boxes"?

GivAlz replied:
I modified a lot the description, as you shall see shortly: I hope it's better...thank you for the comments!

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,35 @@ namespace GridTools
+   *  should be set to a coarse refinement level. If @p refinement_level is higher than the number of levels of the
+   *  triangulation an exception is thrown.
+   * - @p max_boxes : the maximum number of bounding boxes to compute. If more are created the smaller ones are
+   *  merged with neighbors. By default after merging the boxes which can be expressed as a single one no
```
I also don't understand this parameter. I thought boxes are already merged if they "can be expressed as a single one"?

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,35 @@ namespace GridTools
+   *  merged with neighbors. By default after merging the boxes which can be expressed as a single one no
+   *  more boxes are merged. See the BoundingBox::is_neighbor() function for details.
+   *  Notice only neighboring cells are merged (see the @p is_neighbor function in bounding box class): if
+   *  the target number of bounding boxes max_boxes can't be reached by merging neighbors an exception is thrown
```
are you trying to say that the set of boxes is not an exact cover but may contain additional space?

GivAlz replied:
By default, unless one uses the most refined level to create the bounding boxes (in which case these boxes shall occupy the same space as cells), the cover probably contains additional space.
The function isn't meant to describe precisely the area/volume occupied by cells with predicate, but to find a good enough approximation.

tjhei commented on source/base/bounding_box.cc:
```diff
@@ -42,6 +42,86 @@ void BoundingBox<spacedim,Number>::merge_with(const BoundingBox<spacedim,Number>
+
+      // Step 1: testing if the boxes are close enough to intersect
+      for ( unsigned int d=0 ; d<spacedim; ++d )
+        if ( bbox1[0][d] - std::numeric_limits<Number>::epsilon() > bbox2[1][d]
```
this epsilon check is a little weird, shouldn't it be relative? (if the value is larger than one, subtracting eps does nothing)

GivAlz replied:
Yes...this was completely wrong. Thank you for finding it!

tjhei commented on source/grid/grid_tools.cc:
```diff
@@ -2088,6 +2088,159 @@ namespace GridTools
+   const std::function<bool (const typename MeshType::active_cell_iterator &)> &predicate,
+   const unsigned int &refinement_level, const unsigned int &max_boxes)
+  {
+    // Algorithm brief description: beging with creating bounding boxes of all cells at
```
"begin"

tjhei commented on source/grid/grid_tools.cc:
```diff
@@ -2088,6 +2088,159 @@ namespace GridTools
+          {
+            if ( std::find(merged_boxes_idx.begin(),merged_boxes_idx.end(),i) == merged_boxes_idx.end())
+              for (unsigned int j=i+1; j<bounding_boxes.size(); ++j)
+                if ( std::find(merged_boxes_idx.begin(),merged_boxes_idx.end(),j) == merged_boxes_idx.end()
```
This is an expensive O(n^2) algorithm, right? How do you think this will work if we have 100'000 cells or more? This needs to be mentioned in the description of the algorithm.

GivAlz replied:
The idea is that one should choose a coarse refinement_level: for my use case I wasn't thinking of more than 10/20 bounding boxes but, as you point out, this algorithm has a high cost.

I will shortly push a different version where, by default, the merging part is avoided altogether.

tjhei commented on source/grid/grid_tools.cc:
```diff
@@ -2088,6 +2088,159 @@ namespace GridTools
+            for (unsigned int i=0; i<merged_b_boxes.size() && not_removed; ++i)
+              if ( i != min_idx && merged_b_boxes[i].is_neighbor(merged_b_boxes[min_idx]) > 1 )
+                {
+                  merged_b_boxes[i].merge_with(merged_b_boxes[min_idx]);
```
why do you need to merge here again (we already merged everything above or not?)?

GivAlz replied:
This is an extra round: in the first part we merge only if merging doesn't change the space you're describing with the bounding boxes ( for instance in 1d [0,1] and [1,2] can be described with [0,2] without losing anything).

Now that I think of it, this passage in 1d won't do anything. In higher dimension the idea is that, for instance:
if I have the 2d bounding box with low/left and high right points being (0,0),(1,1) and the bounding box (0,1),(eps,1+eps), with eps small, I might want (in my case I want to) merge them to (0,0) (1,1+eps): losing "accuracy" in the spatial description but reducing the number of cells I'm using.

What happens now is that, after this refinement, we might have more bounding boxes than we meant: what the algorithm is does is merging smaller bounding boxes. I want to do this only if the first algorithm didn't reduce enough the number of BBoxes.

tjhei commented on tests/base/bounding_box_3.cc:
```diff
@@ -0,0 +1,235 @@
+
+
+// test for BoundingBox<unsigned int spacedim> which tests the function
+// is_neighbour
```
"neighbor" (many times)

GivAlz replied:
Too many actually...sorry about that

GivAlz commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,184 @@
+                {
+                  if (local_cells[i]->is_locally_owned())
+                    {
+                      if (no_bbox)
```
The code has changed quite a lot from this review...I ended up using the same vertex to begin with (avoiding the if)....

GivAlz commented on source/distributed/grid_tools.cc:
```diff
@@ -0,0 +1,184 @@
+                      if (no_bbox)
+                        {
+                          minp = local_cells[i]->center();
+                          maxp = local_cells[i]->center();
```
It was only to initialize: I'm now using the first vertex of the bbox...
I actually modified this from the already existing function  compute_bounding_box  so, if there's another way of doing it, we might change it as well.

GivAlz commented on include/deal.II/base/bounding_box.h:
```diff
@@ -77,14 +102,23 @@ class BoundingBox
+   *
+   * Return an enumerator of type NeighborType.
+   */
+  NeighborType is_neighbor(const BoundingBox<spacedim,Number> &other_bbox) const;
```
I though a bit about this... there are reasons to change the name, but mainly the idea behind the function is also: 0 means "no intersection", non-zero means intersection (and then the idea is that "intersection" can be classified in different ways).

Otherwise I can propose the name neighbor_kind for instance?

GivAlz commented on include/deal.II/base/bounding_box.h:
```diff
@@ -77,14 +102,23 @@ class BoundingBox
 
   /**
-   * Return true if the point is inside the Bounding Box, false otherwise
+   * Return true if the point is inside the Bounding Box within the numerical error,
```
Ok, maybe "within the numerical error" is not useful but confusing...I put it just to say I'm not checking using "=" but "equal within the numerical error". I realize adding this is only cumbersome: it's a numerical library, everything has an error of some kind to it...

GivAlz commented:
Hi everyone! Sorry for the long period of inactivity but I'm also working with @luca-heltai on another part of my project (which I will add at a certain time).

I wanted to thank you for all the comments, in particular @tjhei who really took a lot of trouble reading it all: sorry for these weeks of silence. I tried to look at all comments (but we're over 100 so maybe something slipped).

The big remaining question, I believe, is the name of is_neighbour... I personally like the name because, if we think of neighbouring as "somehow touching/intersecting" then 0 is false and non-0 is true (even though the type is not strictly speaking bool). If you prefer neighbor_kind or something else this is absolutely not a problem.

I added an image and tried to describe the function as well as possible: I guess it would be useful, once our project is complete, to make a tutorial or something out of it so that at least a study-case is shown.

I still have to add the txt file with the minor change, squash commits etc...but before doing so I think you can take a look. Thank you for all the help!
Giovanni

GivAlz commented on include/deal.II/base/bounding_box.h:
```diff
@@ -24,6 +24,31 @@
+ * - @p not_neighbors : value 0
+ * - @p simple_neighbors : value 1. The boxes intersect with an intersection of dimension at most spacedim - 2
+ * - @p attached_neighbors : value 2. These are neighbors with an intersection of dimension > spacedim - 2
+ * - @p mergeable_neighbors : value 3. Neihgbors which can be expressed with a single Bounding Box, e.g.
```
I hope now is better!

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -24,6 +24,39 @@
+enum class NeighborType
+{
+  /**
+   * value 0: not_neighbours i.e. the intersection is empty
```
you don't need to say what value is there, so i would drop `value x:` here and below. Users shall not see any difference if one day you switch to value 999, 998, etc.

davydden replied:
also remove `_` here and below, the names read fine without it.

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -77,14 +110,22 @@ class BoundingBox
+   *
+   * Return an enumerator of type NeighborType.
+   */
+  NeighborType is_neighbor(const BoundingBox<spacedim,Number> &other_bbox) const;
```
I must admit I also prefer NOT to call it `is_neighbor` is this implies `bool`. Maybe something like `get_neihbor_type()`

GivAlz replied:
Ok, get_neighbor_kind I like ;)

bangerth replied:
If the return type is called `NeighborType`, then the function should be called `get_neighbor_type()`, not `get_neighbor_kind()`. But you can of course also rename the type :-)

davydden replied:
>  then the function should be called get_neighbor_type(), not get_neighbor_kind().

exactly what i suggested above :smile:

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   *  the flag is set to true, make sure to choose wisely a coarse enough @p refinement_level.
+   * - @p max_boxes : the maximum number of bounding boxes to compute. If more are created the smaller ones are
+   *  merged with neighbors. By default after merging the boxes which can be expressed as a single one no
+   *  more boxes are merged. See the BoundingBox::is_neighbor() function for details.
```
update that if the name is changed.

davydden commented on include/deal.II/base/bounding_box.h:
```diff
@@ -77,14 +110,22 @@ class BoundingBox
+   *
+   * Return an enumerator of type NeighborType.
+   */
+  NeighborType get_neighbor_kind (const BoundingBox<spacedim,Number> &other_bbox) const;
```
get_neighbor_type

GivAlz commented:
Ok, squashed everything into a single commit, changed name and added to new/minor my addition.

luca-heltai commented:
I think this is in good shape now. Other changes can be addressed with later PR... but I'll let @tjhei  and @Rombur have a last look.

Rombur commented on source/grid/grid_tools.cc:
```diff
@@ -2088,6 +2088,169 @@ namespace GridTools
+    else
+      {
+        // Merging part of the algorithm
+        // Part 1: merging neighbours
```
Use neighbors

Rombur commented on source/grid/grid_tools.cc:
```diff
@@ -2088,6 +2088,169 @@ namespace GridTools
+        std::vector<unsigned int> merged_boxes_idx;
+        bool found_neighbors = true;
+
+        // We merge only nighbours which can be expressed by a single bounding box
```
typo nighbours

Rombur commented on source/grid/grid_tools.cc:
```diff
@@ -2088,6 +2088,169 @@ namespace GridTools
+        // Part 2: if there are too many bounding boxes, merging smaller boxes
+        // This has sense only in dimension 2 or greater, since  in dimension 1,
+        // neighboring intervals can always be merged without problems
+        if (merged_b_boxes.size() > max_boxes && spacedim > 1)
```
Can you add (), i.e.,write ```if (() && ())```. It makes it easier to read the condition.

GivAlz commented on source/grid/grid_tools.cc:
```diff
@@ -2088,6 +2088,169 @@ namespace GridTools
+      compute_cell_predicate_bounding_box
+      (const typename MeshType::cell_iterator &parent_cell,
+       const std::function<bool (const typename MeshType::active_cell_iterator &)> &predicate,
+       bool &has_predicate)
```
Yes, it probably is better to return tuple, thanks! I'm fixing this and the other typos and I shall update after testing everything!

GivAlz commented:
Added the required modifications. I changed a few things (because it wouldn't compile otherwise):
- added "include tuple" in grid_tools.cc and made the function return a tuple, as suggested
- The function now returns NeighborType:: name of the type (I don't know why it wouldn't compile otherwise) and I had to add (int) to boundary_test_3 so that it would display output
- In the Part 2 of the merging (for compute_mesh_predicate) I now check if the value is "attached" or "merged" and not if value is > 1. I think this way si more readable and, if one day the values are changed, it shall still work.
- Test bounding box was failing because I changed it (adding the std::numeric_limits<Number>::epsilon()...): now it checks that point-bounday > std::numeric_limits<Number>::epsilon()

luca-heltai commented:
@tjhei ?

GivAlz commented:
Ok, now I think point_inside is using properly the epsilon:
`std::numeric_limits<Number>::epsilon()*(std::abs( this->boundary_points.first[i] + p[i]) )
           < this->boundary_points.first[i] - p[i]`
(and the other check is similar).

I'm sorry for this last minutes edits but the first test was failing...

luca-heltai commented:
@tjhei ping?

tjhei commented on doc/news/changes/minor/20171020GiovanniAlzetta:
```diff
@@ -0,0 +1,4 @@
+New: Added a method and a enum class to the class BoundingBox. The method is get_neighbor_type which, given another BoundingBox, returns the enum type NeighborType. This enum class describes how the bounding boxes are neighbor.
```
Please shorten and include the fully qualified name of the class. Maybe something like
"Added BoundingBox::get_neighbor_type which [...]"

tjhei commented on doc/news/changes/minor/20171020GiovanniAlzetta:
```diff
@@ -0,0 +1,4 @@
+New: Added a method and a enum class to the class BoundingBox. The method is get_neighbor_type which, given another BoundingBox, returns the enum type NeighborType. This enum class describes how the bounding boxes are neighbor.
+Added the function compute_mesh_predicate_bounding_box in grid_tools. This function, given a refinement level, returns a vector of BoundingBoxes for which a predicate is true. There are options to reduce the BoundingBox number with and without changing the space's description given by the union of the BoundingBoxes.
```
GridTools::compute_...

tjhei replied:
Note: the reason is that this a) makes it possible to search for the new functions and b) automatically links to them in the doxygen changelog, see http://www.dealii.org/developer/doxygen/deal.II/changes_after_8_5_0.html

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
     const std::function<bool (const typename MeshType::active_cell_iterator &)> &predicate );
 
+  /**
+   * Compute a collection of bounding boxes so that all cells for which the given predicate is true, are
```
very nice summary!

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   * Compute a collection of bounding boxes so that all cells for which the given predicate is true, are
+   * completely enclosed in at least one of the bounding boxes.
+   *
+   * For each cell at a given refinement level containing active cells for which predicate is true,
```
@p predicate

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   * The parameters to control the algorithm are:
+   * - @p predicate : the property of the cells to enclose e.g. IteratorFilters::LocallyOwnedCell .
+   *  The predicate is tested only on active cells.
+   * - @p refinement_level : it defines the level at which the initial bounding box are created. The refinement
```
what happens to active cells that are coarser than this level, are they ignored?

GivAlz replied:
Added

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   * completely enclosed in at least one of the bounding boxes.
+   *
+   * For each cell at a given refinement level containing active cells for which predicate is true,
+   * the function creates a bounding box of its children for which predicate is true.
```
I don't understand. Active cells don't have children.

luca-heltai replied:
he's referring to `cells at a given refinement level`. Those *do* have children in general....

tjhei replied:
I see. So would "For every active cell on a level equal or larger than @p refinement_level and with @p predicate true, the function creates a bounding box." be correct?

GivAlz replied:
Not really... if all children of "cell" (which is not active) have property we just use "cell" and not its children to create the bounding box (this is done by compute_cell_predicate_bounding_box).

tjhei replied:
So it is always a single box for each cell on the given level covering all children with the predicate? Then I totally misunderstood.

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   * For each cell at a given refinement level containing active cells for which predicate is true,
+   * the function creates a bounding box of its children for which predicate is true.
+   *
+   * This results in a cover of these cells; the parameters @p allow_merge and @p max_boxes are used to
```
Please specify exactly what you mean by "these". All active cells with the predicate that have a level >= the one specified?

GivAlz replied:
All active cells for which predicate it's true

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   *  triangulation an exception is thrown.
+   * - @p allow_merge : This flag allows for box merging and, by default, is false. The algorithm has a cost of
+   *  O(N^2) where N is the number of the bounding boxes created from the refinement level; for this reason, if
+   *  the flag is set to true, make sure to choose wisely a coarse enough @p refinement_level.
```
I don't understand why a coarser level gives fewer boxes. Maybe I misunderstand exactly what is going on.

GivAlz replied:
I'm sorry, I believe ffrom your comments I think you slightly misunderstood a part of the algorithm.
The idea is to set a refinement level: this is chosen. Now, for simplicity assume there are no active cells at refinement level or coarser levels. Then there shall be a maximum number of  bounding boxes which is given by the number of cells at this level.
What happen is: call C the a cell at this level. If it has some children which have the property set to true, then you create a box that:
- it's the smallest box inside C and containing all the active cells inside C for which the property is true
- it might possibly contain some more area, unless the geometry is regular

Now you shall have a set of these bounding boxes. You can see that if, instead of C I choose its parent, I shall obtain less Bounding Boxes and less accuracy (without thinking about the merging part).

tjhei replied:
Okay, I think I finally got it.

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   *  If @p max_boxes was left as default or bigger than 1 these two boxes would be returned.
+   * - 4. Because @p max_boxes = 1 the smallest box is merged to the bigger one.
+   * Notice it is important to choose the parameters wisely. For instance, @p allow_merge = false and
+   * @p refinement_level = 1 returns the very same bouding box but with a fraction of the computational cost.
```
typo "bouding box"

GivAlz replied:
Thank you very much @tjhei : I just saw the comments (this weekend I was busy trying to make MPI work) now I shall update everything.

Now that you point it out finding all children of a cell is quicker. I was thinking about something like:
- new_cell = cell-> parent()
- check if all children are "locally owned"
- if so and cell->level < refinement_level then new_cell = new_cell -> parent()
- check all the other children: if all locally owned new_cell = new_cell -> parent
and then I could also do some tricks with merging (following this algorithm you can "avoid" merging one by one all the cells)

The problem is that after this I might still want to run the heavy algorithm because there might still be "mergeable" bounding boxes which are not children of the same cell. Anyway it seems a good improvement, I shall add it asap

tjhei replied:
> I shall add it asap

We could also wait and do it in a followup PR. I am just throwing ideas/comments at you so I can understand things better.

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   *  O(N^2) where N is the number of the bounding boxes created from the refinement level; for this reason, if
+   *  the flag is set to true, make sure to choose wisely a coarse enough @p refinement_level.
+   * - @p max_boxes : the maximum number of bounding boxes to compute. If more are created the smaller ones are
+   *  merged with neighbors. By default after merging the boxes which can be expressed as a single one no
```
The important part of this argument is that you are merging boxes that increase the area (not an exact cover anymore), right? If this is the default value we always get an exact cover?

tjhei replied:
^ If my assumption is correct, please state this.

GivAlz replied:
Unless the geometry is favourable it won't be an exact cover...I guess stating this might be important.

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,52 @@ namespace GridTools
+   *  merged with neighbors. By default after merging the boxes which can be expressed as a single one no
+   *  more boxes are merged. See the BoundingBox::get_neighbor_type () function for details.
+   *  Notice only neighboring cells are merged (see the @p get_neighbor_type  function in bounding box class): if
+   *  the target number of bounding boxes max_boxes can't be reached by merging neighbors an exception is thrown
```
why is it not possible to always take the union of all bounding boxes?

GivAlz replied:
Because usually if b1, b2 are bounding boxes, then b1 union b2 (the union is done thinking them as closed sets in the space) is strictly contained inside b1 merged to b2.
The NeighbourType tries to avoid making a very awful union. For instance
|A|0|
|0|A|
Imagine the A are two bounding boxes (the 0 are owned by someone else): I don't want to merge the two A as it would result into a bounding box of the whole space.

tjhei replied:
okay, so when creating the boxes you are fine with an inexact cover, but when merging you disallow a change of the covered area?

tjhei commented:
I finally understand what you are computing here (thanks for taking the time to explain!). My goal is to get to better and more flexible functionality. Writing good documentation is hard especially when I am nitpicking. Sorry to be such a pain. 

Would you mind explaining to me how you are planning to use this functionality? I am unsure how I could use it in practice because:
1. If my mesh has many coarse cells, the method will generate a very large number of boxes and merging is probably too expensive.
2. If I have a problem with a single coarse cell and I pick the refinement_level=0 each processor has a single bounding box and they are likely to completely overlap.
3. If in 2) I pick a refinement_level>0 I have to be careful because I will miss all cells on coarser levels (correct?).

So, how are you going to pick refinement_level to get a reasonable number of boxes that also cover the selected cells in different situations?

Stepping back from the implementation, ideally we probably want something like "give me a set of about n bounding boxes that best approximate all cells with the predicate", right?

GivAlz commented:
About 3: actually there's a loop on active cells of levels < refinement_level so all that happens is that, for cells in those levels, I shall create a single cell so that's not a problem. And yes, the idea is correct.
As for choosing the refinement_level  I guess one should do some tests and guess it as well as possible: I understand this is not optimal, but it sort of works.

Anyway...I am at the same time very happy and unhappy to announce I had a revolutionizing idea about how to do this: a much quicker algorithm (O(N), in some parts O(NlogN) but with a smaller N) with no need to choose a refinement level.

So where's the catch? I have pretty much to write again everything, but the code should look much better (and I'm very sorry for everyone who approved the code hoping not to see this post again for a while...).

But I may be wrong: I don' think it will take more than a day to write the new version...I'll keep you posted soon ;)

GivAlz commented:
Hi! Today I talked with @luca-heltai about the problem. So here are some answers :

1. Yes, currently the algorithm takes long in a case with lots of coarse cells for which the property is true. But this is not what we want to use it for...and, if the property is true only in a part of the domain, it is still fast
2. This is also a problem, in part. First if there are few (i.e. more than 4/5) processes then it's unlickely that even refinement_level = 0 covers everything. Moreover we shall address the problem with another function which checks for "critical" parts in the bounding box description: if too many overlap then with further communication these (only these) are "refined"
3. As I said yesterday: that's not a problem because currently the first thing the algorithm does is looping over active cells at coarser levels and creating a box from them.

As for the algorithm I was talking about yesterday: there are many ideas...Luca also gave me (and pointed out to the library Rombur linked me which I should study better). The problem is that, in order to take a different approach, it should probably be written again from scratch and then pass again all the verification part.
So... yes this function can clearly be improved but currently (we think) it's good and fast for what we want to do. I believe we could postpone the "make it fast and perfect for all cases" for another PR. Especially because only using it we shall really understand which parts are really flaws, which are not...thank you for all your comments @tjhei :)

tjhei commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -1032,6 +1032,55 @@ namespace GridTools
+   *  The predicate is tested only on active cells.
+   * - @p refinement_level : it defines the level at which the initial bounding box are created. The refinement
+   *  should be set to a coarse refinement level. A bounding box is created for each active cell at coarser
+   *  level than @p refinement_level; ff @p refinement_level is higher than the number of levels of the
```
typo "if"

tjhei commented on doc/news/changes/minor/20171020GiovanniAlzetta:
```diff
@@ -0,0 +1,3 @@
+New: Added BoundingBox::get_neighbor_type() which , given another BoundingBox, returns the enum type NeighborType. Added the function GridTools::compute_mesh_predicate_bounding_box() in grid_tools which returns a vector of BoundingBoxes containing the cells for which a given predicate is true.
```
"which returns the neighboring relation between two bounding boxes. Added the function GridTools::compute_mesh_predicate_bounding_box() which returns a vector of BoundingBoxes covering the cells for which a given predicate is true."

GivAlz commented:
Text changes done, thank you @tjhei :) Currently I need to work on a new part of the project but then I shall have time to implement/think of a better way of doing this.

Thank you also to everyone: my original code was way worse than this!
Best,
Giovanni

tjhei commented:
Okay to merge if the tests pass.

I have some suggestions on how to simplify the logic and avoid the problem of having to select a refinement level in case you are interested.

GivAlz commented:
Hi! I've been thinking about the two tests (Test #2544: bits/step-15.debug and dofs/interpolate_q_system_mixed_01.debug) failing, but honestly I don't know why it's happening: I modified bounding_boxes.h and added a function to grid_tools.h so I don't understand how this affects the two tests...

bangerth commented:
These two tests failed at the point where you branched for this PR. It is not the fault of your changes. We can either ignore this, or you can rebase onto master and the problem will go away.

davydden commented:
@GivAlz 

> you can rebase onto master and the problem will go away.

please rebase, then everything shall be ok.

GivAlz commented:
Ok, thank you for the advice! I just rebased the commit, I hope it shall work now!

jppelteret commented:
@luca-heltai @Rombur @tjhei @davydden Can this be merged?
